File: 프로그래머스/lv3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int answer = 0;
    int max_step = triangle.size();
    
    for(int step = 1; step < max_step; step++) {
        int max_idx = triangle[step].size();
        for(int idx = 0; idx < triangle[step].size(); idx++) {
            if(idx == 0) {
                triangle[step][idx] += triangle[step - 1][idx];
            }
            else if(idx == max_idx - 1) {
                triangle[step][idx] += triangle[step - 1][idx - 1];
            }
            else {
                triangle[step][idx] += max(triangle[step - 1][idx - 1], triangle[step - 1][idx]);
            }
        }
    }
    
    answer = *max_element(triangle[max_step - 1].begin(), triangle[max_step - 1].end());
    
    return answer;
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    dp = [[0 for _ in range(idx + 1)] for idx in range(len(triangle))]
    dp[0][0] = triangle[0][0]
    
    for step in range(1, len(triangle)):
        for idx in range(step + 1):
            if idx == 0:
                dp[step][idx] = dp[step - 1][idx] + triangle[step][idx]
            elif idx == step:
                dp[step][idx] = dp[step - 1][idx - 1] + triangle[step][idx]
            else:
                dp[step][idx] = max(dp[step - 1][idx - 1], dp[step - 1][idx]) + triangle[step][idx]
    
    return max(dp[-1])